canonical/websites-content-system
# [WD-13389] creating a new webpage (without a Jira task)

## Done

 - Added prefix `/api` to all existing endpoints to separate them from other app routes
 - Added a new endpoint `/api/create-page` that create a webpage in the database, sets its reviewers (if any) and generates a copy doc in Google Drive.
 - Created a new route `/new-webpage` that opens a form for creating a new webpage. It can be opened through "Request new page" button on the sidebar. Submitting this form calls `/api/create-page` endpoint and after successful creation of a new webpage navigates to its view.
 - Refactored the way how the tasks work for loading and updating the tree of webpages. Now only loading happens on a schedule basis and update (for new and removed webpage) happens only when a tree is fetched through `/api/get-tree` endpoint.
 - Changed the logic of fetching the tree through `/api/get-tree` endpoint. Before it was fetching it always from the repository itself and database was only used to write and updated tree to it. Now the tree is always fetched from the database. And the cache is also filled with the data from the database. So basically now the source of truth is the database. And the repository is only used to update it (in the background or when the database is initially empty).

## QA

### QA steps

 - Run the app locally through `dotrun`. Make sure you have all the credentials for integration with Google Drive and local database.
 - Click on "Request new page" button from the left sidebar.
 - Fill in the form. Make sure that "Save and generate copy doc" button is disabled until you specify the mandatory fields, which are:
    - URL Title
    - Location
    - Owner
 - Check that the "Save and generate copy doc" button is changed to "Save" if you put a string into "Copy doc" field.
 - Click on the button and check that it correctly navigates to the new page's view after successful creation of the page.
 - Check that the new page is highlighted in the sidebar.
 - Refresh the app and check that the new webpage still shows in the sidebar.

## Fixes

 - Fixes https://warthogs.atlassian.net/browse/WD-13362
 - Fixes https://warthogs.atlassian.net/browse/WD-13389
